src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -930,6 +938,99 @@ void DatabaseCatalog::enqueueDroppedTableCleanup(StorageID table_id, StoragePtr
         (*drop_task)->schedule();
 }
 
+void DatabaseCatalog::dequeueDroppedTableCleanup(StorageID table_id)
+{
+    String latest_metadata_dropped_path;
+    StorageID dropped_table_id = table_id;
+    TablesMarkedAsDropped::iterator table;
+    {
+        std::lock_guard lock(tables_marked_dropped_mutex);
+        table = tables_marked_dropped.end();
+        time_t latest_drop_time = std::numeric_limits<time_t>::min();
+        for (auto it = tables_marked_dropped.begin(); it != tables_marked_dropped.end(); ++it)
+        {
+            if (it->table_id.uuid == table_id.uuid)
+            {
+                table = it;
+                break;
+            }
+
+            if (it->table_id.database_name == table_id.database_name &&
+                it->table_id.table_name == table_id.table_name &&
+                it->drop_time >= latest_drop_time)
+            {
+                latest_drop_time = it->drop_time;
+                table = it;
+            }
+        }
+
+        if (table == tables_marked_dropped.end())
+            throw Exception(ErrorCodes::UNKNOWN_TABLE,
+                "The drop task of table {} is in progress, has been dropped or the database engine doesn't support it",
+                table_id.getNameForLogs());
+
+        latest_metadata_dropped_path = table->metadata_path;
+        dropped_table_id = table->table_id;
+
+        tables_marked_dropped.erase(table);
+        [[maybe_unused]] auto removed = tables_marked_dropped_ids.erase(dropped_table_id.uuid);
+        assert(removed);
+    }
+
+    LOG_INFO(log, "Trying Undrop table {} from {}", dropped_table_id.getNameForLogs(), latest_metadata_dropped_path);
+    String table_metadata_path = getPathForMetadata(dropped_table_id);
```

Comment:
Even if we found the table it may be unsafe to undrop it. The table may contain broken data if `IStorage::drop` was already called, but it has thrown an exception.